src/machine.rs

```diff
@@ -404,6 +412,60 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
         Ok(())
     }
 
+    fn eval_maybe_thread_local_static_const(
+        ecx: &InterpCx<'mir, 'tcx, Self>,
+        mut val: mir::interpret::ConstValue<'tcx>,
+    ) -> InterpResult<'tcx, mir::interpret::ConstValue<'tcx>> {
+        match &mut val {
+            mir::interpret::ConstValue::Scalar(Scalar::Ptr(ptr)) => {
+                let alloc_id = ptr.alloc_id;
+                let alloc = ecx.tcx.alloc_map.lock().get(alloc_id);
+                let tcx = ecx.tcx;
+                let is_thread_local = |def_id| {
+                    tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::THREAD_LOCAL)
+                };
+                match alloc {
+                    Some(GlobalAlloc::Static(def_id)) if is_thread_local(def_id) => {
+                        let new_alloc_id = if let Some(new_alloc_id) =
+                            ecx.get_thread_local_alloc_id(alloc_id)
+                        {
+                            new_alloc_id
```

Comment:
Is this the case "thread-local has already been instantiated for this thread" and the other case is "lazily allocate thread-local for this thread"? If so, please explain that in comments. (Generally there are fairly few comments here considering how subtle the things going on are.)